### DISCUSSION TITLE
Optimism trust assumptions - discussion thread

### DISCUSSION THREAD

**Discussion by panoel** [wormhole-foundation/wormhole#2124] [Protocol / feature design discussion]
Optimism watcher needs to have trust assumptions documented inside the code.

**Comment by panoel** [comment]
When running an Optimism node for the guardian we have to pick one of two choices:

1. Sync from L1
    - Pros
        - Node only gets the L2 blocks that have been submitted to L1
    - Cons
        - Syncing is much slower
2. Sync from L2 - In this mode the node syncing from another Optimism L2 node and may 
    - Pros
        - Faster synchronization
        - Easier to follow head of L2 chain
    - Cons
        - Node doesn't wait for batches to be submitted to L1

The Optimism watcher is using GetLastBlockNumber() and GetTotalElements() to associate a L2 block to an L1 block.  The watcher then waits for the associated L1 block to be finalized before labeling the associated L2 block as finalized. 

When GetTotalElements() returns an L2 index.  That L2 index has already been submitted to L1.

**Comment by aadam-10** [comment]
discussed in https://github.com/wormhole-foundation/wormhole/discussions/2018
